Fix playback of bus encryption enabled bluray discs

cpmw167 opened the pull request:
## Description
Here's my attempt to fix issue https://github.com/xbmc/xbmc/issues/16672

I'm not a programmer and I'm not very familiar with Kodi codebase. Therefore I don't know whether this is the right way of fixing this or not.

Since this is a regression from Krypton and since I can't test whether it breaks menus on SMB protocol, I'll leave it up to Kodi devs to decide whether it should be backported to Leia or not. However, this is currently broken in Leia as well.

## Motivation and Context
Bluray disc playback has been partially broken ever since commit https://github.com/xbmc/xbmc/commit/25d755c0365b2699161bc303bd55afae895fc2ec where Kodi began opening Bluray discs using bd_open_files() instead of bd_open(). According to original commit, this fixes menus when using protocol like SMB, but unfortunately at the same time it breaks playback of all bus encryption enabled bluray discs when using libbluray/libaacs.

Bus encryption is a "protection" that is built in to newer drives, which makes the drives re-encrypt the already encrypted content on-the-fly using something called "data key" (when disc enables BE). Data key is something that drive generates using some secret key and metadata from the disc and it needs to be read from the drive in a special way during AACS authentication (see AACS specifications for details). Then that key must be used for decrypting the bus encryption before the actual content can be decrypted.

When disc is opened using bd_open_files(), libbluray doesn't call aacs_decrypt_bus() which leads in to playback failure because content remains encrypted. This can be fixed by opening discs using bd_open_disc() instead. The old bd_open() which was used in <= Krypton is just a shortcut for bd_open_disc().

I'm attempting to fix this, hopefully without breaking menus on SMB protocol, by adding a special case for SMB where bd_open_files() is used instead of bd_open_disc().

Note that reverting back to bd_open_disc() also enables UHD bluray playback with some restrictions.

## How Has This Been Tested?
I have tested these features on Linux (X11)
- BD folder playback from hard drive (including menus)
- BEE BD playback from disc (including menus)
- UHD BD playback from disc (menus are currently not supported in Kodi)

Note that I can't test BD folder playback on SMB protocol so I can't actually test whether the menus are still working.

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

Commit:
[bluray] fix playback of bus encryption enabled discs

When AACS protected Blu-ray discs are opened using bd_open_files(),
libbluray doesn't call aacs_decrypt_bus(). Bus Encryption Enabled (BEE)
Blu-ray discs must to be opened using bd_open_disc() instead, otherwise
Bus Encryption doesn't get handled properly.

Not handling Bus Encryption will make playback fail when following
conditions are met:
1) drive is Bus Encryption Capable (BEC) and
2) disc enables it (BEE).

Fix this by opening all AACS protected Blu-ray discs with bd_open_disc().

arnova commented on xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamBluray.cpp:
```diff
@@ -217,12 +221,21 @@ bool CDVDInputStreamBluray::Open()
     }
   }
-  else
+  else if (openFiles)
```
Get rid of the openFiles-var and just make this "else if (m_item.IsSmb())".

bahusoid replied:
It seems broken logic affects only blue-ray discs. So shouldn't this be 
```suggestion
  else if (!m_item.IsBluray())
```
instead?

arnova replied:
Yup, makes sense (related to my question below about NFS etc.)

cpmw167 replied:
> It seems broken logic affects only blue-ray discs. So shouldn't this be
> 
> instead?

Yes that is probably better.

arnova commented:
I think this is ok but I'm wondering about one thing: does this only apply to smb? How about nfs/webdav etc. ?

Rechi commented on xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamBluray.cpp:
```diff
@@ -217,12 +221,21 @@ bool CDVDInputStreamBluray::Open()
     if (!bd_open_files(m_bd, &m_rootPath, CBlurayCallback::dir_open, CBlurayCallback::file_open))
     {
-      CLog::Log(LOGERROR, "CDVDInputStreamBluray::Open - failed to open %s", CURL::GetRedacted(root).c_str());
+      CLog::Log(LOGERROR, "CDVDInputStreamBluray::Open - failed to open %s in files mode", CURL::GetRedacted(root).c_str());
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamBluray.cpp:
```diff
@@ -217,12 +221,21 @@ bool CDVDInputStreamBluray::Open()
+  else
+  {
+    m_rootPath = root;
+    if (!bd_open_disc(m_bd, CURL::GetRedacted(root).c_str(), NULL))
```
> Prefer the use of `nullptr` instead of `NULL`. [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#123-null-vs-nullptr

Rechi commented on xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamBluray.cpp:
```diff
@@ -217,12 +221,21 @@ bool CDVDInputStreamBluray::Open()
+    m_rootPath = root;
+    if (!bd_open_disc(m_bd, CURL::GetRedacted(root).c_str(), NULL))
+    {
+      CLog::Log(LOGERROR, "CDVDInputStreamBluray::Open - failed to open %s in disc mode", CURL::GetRedacted(root).c_str());
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi reviewed (commented):
Please follow the [current code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

cpmw167 commented:
> I think this is ok but I'm wondering about one thing: does this only apply to smb? How about nfs/webdav etc. ?

That is a good question. I wish I knew.

I would guess that in the worst case this breaks menu playback on these protocols, but playback without menus should still work though.

cpmw167 commented:
I've updated this according to the comments above.

All my test cases are still working. I just don't know whether the menus are still working on protocols like nfs, smb, webdav etc.

arnova commented:
Thanks for updating it. Only thing missing is a comment why this is needed, as soon as that's in place I think it's good to go as is.

bahusoid commented on xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamBluray.cpp:
```diff
@@ -217,12 +217,23 @@ bool CDVDInputStreamBluray::Open()
+  else
+  {
+    m_rootPath = root;
+    if (!bd_open_disc(m_bd, CURL::GetRedacted(root).c_str(), nullptr))
```
Why `GetRedacted` call is used here. It wasn't used before https://github.com/xbmc/xbmc/commit/25d755c0365b2699161bc303bd55afae895fc2ec#diff-276823fbbb9dfcd26bb50ffee871a7c3L383

cpmw167 replied:
> Why `GetRedacted` call is used here. It wasn't used before [25d755c#diff-276823fbbb9dfcd26bb50ffee871a7c3L383](https://github.com/xbmc/xbmc/commit/25d755c0365b2699161bc303bd55afae895fc2ec#diff-276823fbbb9dfcd26bb50ffee871a7c3L383)
> It seems this method should be used only for logging purposes.

No reason, I didn't know it's only used for logging purposes. When I first experimented with this I just copy pasted it and forgat about it. But you are right, it isn't necessary and I've dropped it now.

cpmw167 commented:
> Thanks for updating it. Only thing missing is a comment why this is needed, as soon as that's in place I think it's good to go as is.

I've added a commit message with a reasoning why this is needed. I hope this is sufficient?

arnova commented:
> > Thanks for updating it. Only thing missing is a comment why this is needed, as soon as that's in place I think it's good to go as is.
> 
> I've added a commit message with a reasoning why this is needed. I hope this is sufficient?

No ;-) : Please also add a comment in the code itself so future "readers" also know why this is special cased.

cpmw167 commented:
> No ;-) : Please also add a comment in the code itself so future "readers" also know why this is special cased.

Done :)

lharms commented:
I tried this and it caused ISO BR to fail to load on SMB shares picking individual streams.  The bluray menu option did work however over SMB.

2020-01-08 20:40:19.627 T:7496  NOTICE: VideoPlayer::OpenFile: bluray://udf%3a%2f%2fsmb%253a%252f%252fDISKSTATION%252fmv%252fMOVIES%252fZombieland%2520(2009)%252fZombieland%2520(2009)%2520bluray.iso%2f/BDMV/PLAYLIST/00001.mpls
2020-01-08 20:40:19.631 T:6548  NOTICE: Creating InputStream
2020-01-08 20:40:19.710 T:6548   ERROR: CDVDInputStreamBluray::Open - failed to open udf://smb%3a%2f%2fDISKSTATION%2fmv%2fMOVIES%2fZombieland%20(2009)%2fZombieland%20(2009)%20bluray.iso/ in disc mode
2020-01-08 20:40:19.712 T:6548   ERROR: CVideoPlayer::OpenInputStream - error opening [bluray://udf%3a%2f%2fsmb%253a%252f%252fDISKSTATION%252fmv%252fMOVIES%252fZombieland%2520(2009)%252fZombieland%2520(2009)%2520bluray.iso%2f/BDMV/PLAYLIST/00001.mpls]
2

cpmw167 commented:
> I tried this and it caused ISO BR to fail to load on SMB shares picking individual streams. The bluray menu option did work however over SMB.
> 
> 2020-01-08 20:40:19.627 T:7496 NOTICE: VideoPlayer::OpenFile: bluray://udf%3a%2f%2fsmb%253a%252f%252fDISKSTATION%252fmv%252fMOVIES%252fZombieland%2520(2009)%252fZombieland%2520(2009)%2520bluray.iso%2f/BDMV/PLAYLIST/00001.mpls
> 2020-01-08 20:40:19.631 T:6548 NOTICE: Creating InputStream
> 2020-01-08 20:40:19.710 T:6548 ERROR: CDVDInputStreamBluray::Open - failed to open udf://smb%3a%2f%2fDISKSTATION%2fmv%2fMOVIES%2fZombieland%20(2009)%2fZombieland%20(2009)%20bluray.iso/ in disc mode
> 2020-01-08 20:40:19.712 T:6548 ERROR: CVideoPlayer::OpenInputStream - error opening [bluray://udf%3a%2f%2fsmb%253a%252f%252fDISKSTATION%252fmv%252fMOVIES%252fZombieland%2520(2009)%252fZombieland%2520(2009)%2520bluray.iso%2f/BDMV/PLAYLIST/00001.mpls]
> 2

Strange. It seems `item.IsDiscImage()` isn't working as intended when ISO image is loaded using smb:// protocol.

My patch doesn't touch iso image playback at all so it should work just as well as it did before. Hence I doubt it was working as intended even before my patch.

AFAICT the code should use bd_open_stream() when disc image is detected. For some reason it's skipping bd_open_stream() and bd_open_files() and then falling back to bd_open_disc().

arnova commented:
I see no reason by IsDiscImage() wouldn't work properly for smb:// @lharms : Please provide a full debug log

cpmw167 commented:
I've changed the logic slightly, I hope this is less "invasive" and I hope I got it right this time.

Basically I've added a new check for protected BDs. When a protected disc is found (i.e. AACS/Unit_Key_RO.inf file is present), disc is opened using bd_open_disc() otherwise it uses the old behaviour. AACS folder shouldn't be present on decrypted BD folders or ISO images.

@lharms Could you perhaps test whether the ISO images are still working on smb:// protocol (with and without menus)?

lharms commented:
Ah clever!  That worked just fine.  

It does rely on that one file missing.  Pretty sure the prog I use does that all the time.  But I could not 100% guarantee that.  

Not sure this is needed but maybe instead of just returning false on the error path of of bd_open_disc go ahead and call the other one bd_open_files?  That way you could still play encrypted streams?  Kinda of gross with the same code in there twice though .  Not sure the variables are in the right way to pull that off.  I am good with the way it is now though.

cpmw167 commented:
> It does rely on that one file missing. Pretty sure the prog I use does that all the time. But I could not 100% guarantee that.

The whole AACS folder is always missing when the source is decrypted. All players assume source is encrypted when AACS folder is present, therefore rippers either drop it or rename it (e.g. AnyDVD renames it to ANY, DVDFab renames it to FAB etc).

> Not sure this is needed but maybe instead of just returning false on the error path of of bd_open_disc go ahead and call the other one bd_open_files? That way you could still play encrypted streams? Kinda of gross with the same code in there twice though . Not sure the variables are in the right way to pull that off. I am good with the way it is now though.

That would be uglier IMHO. Also since this is already quite fragile, why use hacks?

I'm not sure what you mean by encrypted streams. Libbluray/libaacs will always decrypt if necessary, this patch doesn't change that. It's just that bus encryption isn't handled unless bd_open_disc is being used. Plus nobody should do encrypted backups. That's just wrong, because it's technically impossible to derive a decryption key from a backup (missing VID). Also, bd_open_disc works for hard disk folders too in case somebody wants to play an encrypted backup.

Variables are the right way to do this, because we need to do multiple checks and protocol needs to be omitted. Depending on the source we're opening, kodi might use uris like `<path>`, `bluray://<path>`¸ or `bluray://udf://<path>` etc. However, if somebody has a better idea that how to do this without rewriting the whole thing, I'm all ears. Otherwise I would just push the button. :)

lharms commented:
By vars I meant if you wanted to call both twice, the double call idea I
suggested.  Not sure with the way it is written the vars would enable that
in a clean eay.  Sorry did not mean to confuse the issue.  What you have is
fine.  The only check i would say is left, is trying to play an individual
encrypted file instead of the whole disc from the physical disc.  You may
have done that already?

On Fri, Jan 10, 2020, 10:20 AM cpmw167 <notifications@github.com> wrote:

> It does rely on that one file missing. Pretty sure the prog I use does
> that all the time. But I could not 100% guarantee that.
>
> The whole AACS folder is always missing when the source is decrypted. All
> players assume source is encrypted when AACS folder is present, therefore
> rippers either drop it or rename it (e.g. AnyDVD renames it to ANY, DVDFab
> renames it to FAB etc).
>
> Not sure this is needed but maybe instead of just returning false on the
> error path of of bd_open_disc go ahead and call the other one
> bd_open_files? That way you could still play encrypted streams? Kinda of
> gross with the same code in there twice though . Not sure the variables are
> in the right way to pull that off. I am good with the way it is now though.
>
> That would be uglier IMHO. Also since this is already quite fragile, why
> use hacks?
>
> I'm not sure what you mean by encrypted streams. Libbluray/libaacs will
> always decrypt if necessary, this patch doesn't change that. It's just that
> bus encryption isn't handled unless bd_open_disc is being used. Plus nobody
> should do encrypted backups. That's just wrong, because it's technically
> impossible to derive a decryption key from a backup (missing VID). Also,
> bd_open_disc works for hard disk folders too in case somebody wants to play
> an encrypted backup.
>
> Variables are the right way to do this, because we need to do multiple
> checks and protocol needs to be omitted. Depending on the source we're
> opening, kodi might use uris like <path>, bluray://<path>¸ or
> bluray://udf://<path> etc. However, if somebody has a better idea that
> how to do this without rewriting the whole thing, I'm all ears. Otherwise I
> would just push the button. :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/xbmc/xbmc/pull/17145?email_source=notifications&email_token=ACJKHHT6OLI24MA2QPOMLTTQ5CG5DA5CNFSM4KDK5UO2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIUHO5I#issuecomment-573077365>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACJKHHXA4VPU7EYFO5EJLC3Q5CG5DANCNFSM4KDK5UOQ>
> .
>

cpmw167 commented:
> By vars I meant if you wanted to call both twice, the double call idea I suggested. Not sure with the way it is written the vars would enable that in a clean eay. Sorry did not mean to confuse the issue. What you have is fine. The only check i would say is left, is trying to play an individual encrypted file instead of the whole disc from the physical disc. You may have done that already?

I've thought about that double call idea, but since this seems to be very fragile there is no way to know what use case it might break. I think it's better to detect what we're opening and use a right call instead of trying every call be "brute force" until one works.

The open_files vs open_disc thing is not about opening individual files. It's about making sure that AACS authentication is done properly when original disc is opened.

I've tested this by selecting individual playlist from simple menu, selecting full menus, selecting index.bdmv and selecting MovieObject.bdmv and all of them are working. But if you mean playback by selecting individual m2ts files then yes that won't work and probably never has (at least not on BEE disc). That is what the playlists are for. :)

cpmw167 commented:
> By vars I meant if you wanted to call both twice, the double call idea I suggested. Not sure with the way it is written the vars would enable that in a clean eay. Sorry did not mean to confuse the issue. What you have is fine. The only check i would say is left, is trying to play an individual encrypted file instead of the whole disc from the physical disc. You may have done that already?

I think I know what you are asking. I think you have misunderstood what this PR actually does.

These `bd_open_*` functions are not about playing individual streams or "full discs", these are just for populating the information structures. For example these functions will parse and sort playlists and try to guess the main title. These also initialize AACS etc.

All these different calls just tell libbluray what the source is how it should be handled, e.g. in stream mode libbluray assumes the source is a "raw" unmounted disk image or block device and it parses filesystem (udf) structures and reads in raw blocks, where in disc mode it assumes the source is an optical device and then handles things like bus encryption properly. In files mode it assumes we're dealing with a folder on a hard drive, which don't need to be read in "raw" mode and things like bus encryption are not present. All these modes decrypt if necessary (and when possible).

All stream selection is done with `bd_select_(title|stream)` and all that is unrelated to this PR. In short, this PR is about making sure AACS authentication is handled properly and nothing else.

I hope this clears it up.

@arnova Unless there are no objections, could you perhaps merge this?

arnova reviewed (dismissed):
> > By vars I meant if you wanted to call both twice, the double call idea I suggested. Not sure with the way it is written the vars would enable that in a clean eay. Sorry did not mean to confuse the issue. What you have is fine. The only check i would say is left, is trying to play an individual encrypted file instead of the whole disc from the physical disc. You may have done that already?
> 
> I think I know what you are asking. I think you have misunderstood what this PR actually does.
> 
> These `bd_open_*` functions are not about playing individual streams or "full discs", these are just for populating the information structures. For example these functions will parse and sort playlists and try to guess the main title. These also initialize AACS etc.
> 
> All these different calls just tell libbluray what the source is how it should be handled, e.g. in stream mode libbluray assumes the source is a "raw" unmounted disk image or block device and it parses filesystem (udf) structures and reads in raw blocks, where in disc mode it assumes the source is an optical device and then handles things like bus encryption properly. In files mode it assumes we're dealing with a folder on a hard drive, which don't need to be read in "raw" mode and things like bus encryption are not present. All these modes decrypt if necessary (and when possible).
> 
> All stream selection is done with `bd_select_(title|stream)` and all that is unrelated to this PR. In short, this PR is about making sure AACS authentication is handled properly and nothing else.
> 
> I hope this clears it up.
> 
> @arnova Unless there are no objections, could you perhaps merge this?

Please address my code suggestions first about not using a cfileitem for the new function but a std::string (did you miss those?)

cpmw167 commented on xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamBluray.cpp:
```diff
@@ -147,6 +148,11 @@ bool CDVDInputStreamBluray::Open()
+    CURL url3(root);
+    CFileItem base(url3, false);
+    openDisc = base.IsProtectedBlurayDisc();
+
```
Sorry I must have missed this.

I don't know this code very well so my apologies if I'm asking dumb things here, but how would this be better since the m_item check below is already a FileItem, therefore IsProtectedBluray should be in CFileItem, no?

Also, I copied this logic from IsBluray/openDisc checks so why would we want this to be implemented differently than those two? Surely this isn't anymore overkill than those two, or am I missing something here? I mean IsBluray uses the exact same method for checking whether BDMV/index.bdmv exists (for example).

cpmw167 replied:
I meant openStream checks. I mean IsBluray and IsDiscImage are both fileitems and I'm reusing the "item" and "m_item" which are fileitems already, so wouldn't moving IsProtectedBlurayDisc to URIUtils make this more inconsistent and complicated?

cpmw167 commented:
> Please address my code suggestions first about not using a cfileitem for the new function but a std::string (did you miss those?)

@arnova Could you please elaborate on why you want to me to move `IsProtectedBlurayDisc()` to URIUtils? CFileItem is full of similar checks and since `IsProtectedBlurayDisc()` isn't parsing or doing any work on URIs or strings, this doesn't make any sense to me. It's checking whether file is physically present on the source, just like `IsBluray()` for example.

`IsProtectedBlurayDisc()` is working with CFileItems just like `IsDiscImage()` or `IsBluray()` or `IsRAR()` etc. IMHO this is how it should work when `item` is CFileItem

```
bool a = false;
bool b = false;
bool c = false;

a = item.IsBluray();
b = item.IsDiscImage();
c = item.IsProtectedBlurayDisc();
```

Why should that last one be in URIUtils since it's not working with URIs while the first two are in CFileItem?

Hence what you're asking me to do is essentially the same thing as making `IsRAR()` a CFileItem while making `IsZIP()` a URIUtil.

arnova commented:
> > Please address my code suggestions first about not using a cfileitem for the new function but a std::string (did you miss those?)
> 
> @arnova Could you please elaborate on why you want to me to move `IsProtectedBlurayDisc()` to URIUtils? CFileItem is full of similar checks and since `IsProtectedBlurayDisc()` isn't parsing or doing any work on URIs or strings, this doesn't make any sense to me. It's checking whether file is physically present on the source, just like `IsBluray()` for example.
> 
> `IsProtectedBlurayDisc()` is working with CFileItems just like `IsDiscImage()` or `IsBluray()` or `IsRAR()` etc. IMHO this is how it should work when `item` is CFileItem
> 
> ```
> bool a = false;
> bool b = false;
> bool c = false;
> 
> a = item.IsBluray();
> b = item.IsDiscImage();
> c = item.IsProtectedBlurayDisc();
> ```
> 
> Why should that last one be in URIUtils since it's not working with URIs while the first two are in CFileItem?
> 
> Hence what you're asking me to do is essentially the same thing as making `IsRAR()` a CFileItem while making `IsZIP()` a URIUtil.

Ok, agreed.

arnova commented:
TVOS build error unrelated

smf007 commented:
Can this get backported to V18 Leia?  [Issue 16672](https://github.com/xbmc/xbmc/issues/16672) was with v18 Leia.
